lxd/images.go

```diff
@@ -646,15 +668,22 @@ func imageCreateInPool(d *Daemon, info *api.Image, storagePool string) error {
 
 func imagesPost(d *Daemon, r *http.Request) response.Response {
 	trusted, _, _, _ := d.Authenticate(r)
-	if !trusted {
-		secret := r.Header.Get("X-LXD-secret")
-		fingerprint := r.Header.Get("X-LXD-fingerprint")
-
-		if secret != "" && fingerprint != "" {
-			valid := imageValidSecret(fingerprint, secret)
-			if !valid {
-				return response.Forbidden(nil)
-			}
+
+	secret := r.Header.Get("X-LXD-secret")
+	fingerprint := r.Header.Get("X-LXD-fingerprint")
+	var imageMetadata map[string]interface{}
+
+	if secret == "" || fingerprint == "" {
+		if !trusted {
+			return response.Forbidden(nil)
+		}
+	} else {
+		// We need to invalidate the secret whether the source is trusted or not.
+		op, valid := imageValidSecret(fingerprint, secret)
+		if valid {
+			imageMetadata = op.Metadata()
+		} else if !trusted {
+			return response.Forbidden(nil)
 		}
```

Comment:
Ah, so that's the fix for the security bug from earlier commits, can you please rebase the commits so that they're all good when looking at the individually?

Reply:
Wouldn't:

```go
if !trusted {
    secret := r.Header.Get("X-LXD-secret")
    fingerprint := r.Header.Get("X-LXD-fingerprint")
    if secret == "" || fingerprint == "" {
        return response.Forbidden(nil)
    }

    op, valid := imageValidSecret(fingerprint, secret)
    if !valid {
        return response.Forbidden(nil)
    }

    imageMetadata = op.Metadata()
}
```

Be more readable and run less code in the authenticated case?

Reply:
No, we shouldn't do this. Let's assume the source is actually trusted by the target. In that case, the token operation would never be cancelled. We should make sure, the token operation is always cancelled.

Reply:
Hmm, I see. I'd probably argue we don't need a token at all in that case, but I guess that's fine.